opm/output/eclipse/WindowedArray.hpp

```diff
@@ -0,0 +1,177 @@
+/*
+  Copyright (c) 2018 Statoil ASA
+
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_WINDOWED_VECTOR_HPP
+#define OPM_WINDOWED_VECTOR_HPP
+
+#include <cassert>
+#include <iterator>
+#include <type_traits>
+#include <vector>
+
+#include <boost/range/iterator_range.hpp>
+
+namespace Opm { namespace RestartIO { namespace Helpers {
+
+    template <typename T>
+    class WindowedArray
+    {
+    private:
+        using Vector = std::vector<T>;
+
+    public:
+        using WriteWindow = boost::iterator_range<
+            typename Vector::iterator>;
+
+        using ReadWindow = boost::iterator_range<
+            typename Vector::const_iterator>;
+
+        using Idx = typename Vector::size_type;
+
+        struct NumWindows { Idx value; };
+        struct WindowSize { Idx value; };
+
+        explicit WindowedArray(const NumWindows n, const WindowSize sz)
+            : x_         (n.value * sz.value)
+            , windowSize_(sz.value)
+        {}
+
+        Idx numWindows() const
+        {
+            return this->x_.size() / this->windowSize_;
+        }
+
+        Idx windowSize() const
+        {
+            return this->windowSize_;
+        }
+
+        WriteWindow operator[](const Idx window)
+        {
+            assert ((window < this->numWindows()) &&
+                    "Window ID Out of Bounds");
+
+            auto b = std::begin(this->x_) + window*this->windowSize_;
+            auto e = b + this->windowSize_;
+
+            return { b, e };
+        }
+
+        ReadWindow operator[](const Idx window) const
+        {
+            assert ((window < this->numWindows()) &&
+                    "Window ID Out of Bounds");
+
+            auto b = std::begin(this->x_) + window*this->windowSize_;
+            auto e = b + this->windowSize_;
+
+            return { b, e };
+        }
+
+        const Vector& data() const
+        {
+            return this->x_;
+        }
+
+        Vector getDataDestructively()
```

Comment:
This is above my competence when it comes to move semantics and+++; but can you guarantee that calling this method will act destructively? I guess it hinges on the fact that `std::vector<T>` has a move assignment operator implemented - and that is impossible to tell from this method.

Reply:
> move assignment operator implemented

Move construction.  Guaranteed since C++11.